File: baekjoon/silver2/6236.py
# input n, m
# n is # of day, m is # of withdraw
n, m = map(int, input().split())

# list of money to spend
money = []
total = 0
for i in range(n):
  a = int(input())
  money.append(a)
  total += a

# binary search
left = 1
right = total
min_k = 0 # minimum k

while (left <= right):
  mid = (left + right) // 2
  current_money = mid # available money
  check = True # check ( today_money < current_money )
  cnt = 1 # # of withdraw

  for value in money:
    # check1. is it possible? (today_money < current_money ?)
    if(value > mid):
      check = False
      break

    # check2. have enough money?
    if(value > current_money):
      current_money = mid
      cnt += 1

    current_money -= value

  # update left and right
  # case1. (today_money > withdraw_money) or (count > M) ?
  if (not check) or (cnt > m):
    left = mid+1

  # otherwise
  else:
    min_k = mid
    right = mid-1

# finish binary search ( = get minimum k )
print(min_k)

File: baekjoon/silver2/6236.cc
#include <iostream>
using namespace std;

// N is day, M is # of withdraw
int N, M;

// list of money to spend
int money[100001] = {0, };

// get minimum K with binary search
int get_min_money(int total) {
  int result = 0; // minimum K
  int left = 1;
  int right = total;
  int mid = 0; // withdraw money
  
  while(left <= right) {
    mid = (left + right) / 2;
    int current_money = mid;
    int cnt = 1; // # of withdraw
    bool pass = true; // today_money < withdraw money ?
    
    for(int i = 0; i < N; i++) {
      // if today_money > withdraw money? increase withdraw money
      if(money[i] > mid) {
        pass = false;
        break;
      }
      
      // if today_money > current_money? withdraw money
      if(money[i] > current_money) {
        current_money = mid;
        cnt++;
      }

      // spend money
      current_money -= money[i]; 
    }

    // handle withdraw money
    // case1. (today_money > withdraw money) or (# of withdraw > N)
    // increase withdraw money and search again
    if(!pass || cnt > M) {
      left = mid + 1;
    }

    // case2. not case1
    // decrease withdraw money and search again
    else {
      result = mid;
      right = mid - 1;
    }
  }

  return result;
}

int main() {
  
  // input N and M
  cin >> N >> M;
  
  int total_money = 0;
  // input money to spent
  for(int i = 0; i < N; i++) {
    cin >> money[i];
    total_money += money[i];
  }
  
  // get minimum # of K
  int K = get_min_money(total_money);

  // print result K
  cout << K << endl;

  return 0;
}

